### ISSUE TITLE
Trick to double the training speed (improvement)

### ISSUE THREAD

**Issue by zungam** [keras-team/keras-contrib#255]
Hi, I managed to double the training speed of Mask-RCNN with ResNet101 as backbone with a simple change to Keras. Today Keras can use generators to yield batches ready for the next training step. This is good, because it means while the GPU is training, the CPU can make batches ready. However, the problem is if batches is of different sizes, or requires different loading times due to augmentation or other scenerios. If the generator ever takes longer time than GPU, the GPU halts, waiting for the generator to yield results. This can really halt the training time sometimes.

To make the GPU allways have an image ready to load into GPU memory without waiting time, I propose having a FIFO queue fetching batches non-stop (as long as there are memory). This means that while the GPU is busy doing what it does, the CPU will always work in the background to have images ready (forever). This litle trick doubled the training time of my network. I did the following (Only works on windows), so if any cares to implement a real Keras thread to do the same, I would advice for a reimplementation and merge request on it.

**How to:**  
In [keras/engine/training_generator.py](https://github.com/keras-team/keras/blob/master/keras/engine/training_generator.py), add this class:

```
import threading
class ThreadingBatches(object):
    """ Threading example class
    The run() method will be started and it will run in the background
    until the application exits.
    """
    def __init__(self, generator=None):
        """ Constructor
        :type interval: int
        :param interval: Check interval, in seconds
        """
        self.generator = generator
        self.stack = []
        self.queue_size = 50 #This number should be experimented with based on how much RAM you have
        thread = threading.Thread(target=self.run, args=())
        thread.daemon = True                    # Daemonize thread
        thread.start()                                  # Start the execution
        
        

    def run(self):
        """ Method that runs forever """
        while True:
            # Do something
            if(len(self.stack) < self.queue_size):
                try:
                    batch = next(self.generator)
                    self.stack.append(batch)
                except ValueError:
                    None
    def popNextBatch(self):
        while len(self.stack) == 0:
            None

        return self.stack.pop(0)
```

Now, add this below the line 145:
`batch_thread = ThreadingBatches(generator=output_generator) `
and overwrite the line 155 from:
`generator_output = next(output_generator)`
to
`generator_output = batch_thread.popNextBatch()`

And voala, superfast training! :)

The trick should be optional, and queue_size should be customizable.
